snapshot-service/pb/service.proto

```diff
@@ -0,0 +1,265 @@
+
+message Id {
+  string id = 1;
+  uint64 version = 2;
+}
+
+message IdFilter {
+  string id = 1;
+  oneof version {
+    uint64 value = 2;
+    bool all = 3;
+    bool max = 4;
+  }
+}
+
+message Condition {
+  Id id = 1;
+  string name = 2;
+  bool enabled = 3;
+  repeated string device_id_filter = 5;
+  repeated string resource_type_filter = 6;
+  repeated string resource_href_filter = 7;
+  string yq_expression = 8;
+  
+  message InvokeConfiguration {
+    string id = 1; // latest version of cfg
+    bool keep_updating_on_failure = 2;
+    string api_access_token = 3; // token used to update resources in the configuration
+    // or do we want to use token from device to update self resources ?
+  }
+  InvokeConfiguration invoke_configuration = 9;
+}
+
+message CreateConditionReponse { Id id = 1; }
+message UpdateConditionResponse {
+  Id id = 1; // for new version 
+} 
+message GetConditionsRequest { repeated IdFilter id_filter = 1; }
+message DeleteConditionsRequest { repeated IdFilter id_filter = 1; }
+message DeleteConditionsResponse { int64 count = 1; }
+
+
+message ResourceTypes {
+  repeated string types = 1; // all types must match resource
+  uint32 min = 2; // minimal number of resources that will be updated
+}
+
+message Configuration {
+  message Resource {
+    oneof filter {
+      string href  = 1;
+      ResourceTypes resource_types = 2; 
+    }
+    resourceaggregate.pb.Content content = 3;
+    string resource_interface = 6; // optional update interface
+    int64 time_to_live = 7; // optional update command time to live
+  }
+  Id id = 1;
+  string name = 2;
+  repeated Resource resources = 3;
+}
+
+message CreateConfigurationReponse { Id id = 1; }
+message UpdateConfigurationReponse  {
+  Id id = 1; // for new version 
+} 
+message GetConfigurationsRequest { repeated IdFilter id_filter = 1; }
+message DeleteConfigurationsRequest { repeated IdFilter id_filter = 1; }
+message DeleteConfigurationsResponse { int64 count = 1; }
+
+message AppliedConfiguration {
+  message Resource {
+    resourceaggregate.pb.ResourceId resource_id = 1;
+    uint32 configuration_resources_idx = 2; // index of resource in configuration resources array. For resource types it could be mutliple resources.
+    enum Status {
+      QUEUED = 0;
+      INPROGRESS = 1;
+      WAITING_FOR_REOURCE = 2;
+      DONE = 3;
+      TIMEOUT = 4;
+      FAIL = 5;
+    };
+    Status status = 3;
+    int64 timestamp_start = 4; // when the rule association was applied
+    int64 valid_until = 5; // how long the command is valid
+    resourceaggregate.pb.ResourceUpdated resource_updated = 6;
+  }
+  string id = 1;
+  string device_id = 2;
+  Id configuration_id = 3;
+  oneof executed_by {
+    bool on_demaned = 4;
+    Id condition_id = 5;
+  }
+  repeated Resource resources = 6;
+}
+
+message InvokeConfigurationRequest {
+  string id = 1;
+  repeated string device_id_filter = 2; // at least one must be set
+  bool force = 3;
+  bool keep_updating_on_failure = 4;
```

Comment:
@jkralik could you describe scenarios for both, force and keep_updating please?

Reply:
- force: re-apply configuration to device
- keep_updating_on_failure: when some update resource fails the updates will continue (maybe this flag could be moved to Resource Config so for each update resource you can said if you want to continue)